core/ledger/kvledger/snapshot_mgmt.go

```diff
@@ -0,0 +1,342 @@
+		return err
+	}
+	if exists {
+		return errors.Errorf("duplicate snapshot request for height %d", height)
+	}
+
+	if err = l.snapshotRequestBookkeeper.add(height); err != nil {
+		return err
+	}
+
+	if height == bcInfo.Height {
+		keepLock = true
+		go func() {
+			if err := l.generateSnapshot(); err != nil {
+				logger.Errorw("failed to generate snapshot", "height", height, "error", err)
+			}
+		}()
+	}
+
+	return nil
+}
+
+// CancelSnapshotRequest cancels the previously submitted request.
+// It returns an error if such a request does not exist or is under processing.
+// It locks snapshotRequestLock to synchronize among concurrent SubmitSnapshotRequest
+// and CancelSnapshotRequest calls.
+// It also locks snapshotGenerationRWLock to synchronize with commit before getting
+// blockchain info so that it will not cancel a under-processing request.
+func (l *kvLedger) CancelSnapshotRequest(height uint64) error {
+	l.snapshotRequestLock.Lock()
+	defer l.snapshotRequestLock.Unlock()
+
+	l.snapshotGenerationRWLock.RLock()
+	defer l.snapshotGenerationRWLock.RUnlock()
+
+	exists, err := l.snapshotRequestBookkeeper.exists(height)
+	if err != nil {
+		return err
+	}
+	if !exists {
+		return errors.Errorf("no pending snapshot request has height %d", height)
+	}
+
+	bcInfo, err := l.blockStore.GetBlockchainInfo()
+	if err != nil {
+		return err
+	}
+	if height == bcInfo.Height {
+		return errors.Errorf("cannot cancel the snapshot request because it is under processing")
+	}
+
+	return l.snapshotRequestBookkeeper.delete(height)
+}
+
+// PendingSnapshotRequests returns a list of heights for the pending (or under processing) snapshot requests.
+func (l *kvLedger) PendingSnapshotRequests() ([]uint64, error) {
+	return l.snapshotRequestBookkeeper.list()
+}
+
+// ListSnapshots returns the information for available snapshots.
+// It returns a list of strings representing the following JSON object:
+// type snapshotSignableMetadata struct {
+//    ChannelName        string            `json:"channel_name"`
+//    ChannelHeight      uint64            `json:"channel_height"`
+//    LastBlockHashInHex string            `json:"last_block_hash"`
+//    FilesAndHashes     map[string]string `json:"snapshot_files_raw_hashes"`
+// }
+func (l *kvLedger) ListSnapshots() ([]string, error) {
+	return nil, errors.Errorf("not implemented")
+}
+
+// DeleteSnapshot deletes the snapshot files except for the metadata file and
+// returns an error if no such a snapshot exists
+func (l *kvLedger) DeleteSnapshot(height uint64) error {
+	return errors.Errorf("not implemented")
+}
+
+// generateSnapshotIfNeeded generates a snapshot if the block height matches smallest request height.
+// It locks snapshotGenerationRWLock to synchronize between commit and snapshot generation.
+// It starts a go-routine to call generateSnapshot if needed and will keep the snapshotGenerationRWLock
+// until generateSnapshot is finished.
+func (l *kvLedger) generateSnapshotIfNeeded() error {
```

Comment:
You need to take `snapshotRequestLock` here, else, following are the two side effects
- Access to the variable `l.snapshotRequestBookkeeper.smallestRequestHeight` in this function is unprotected
- This function can can concurrently execute with the function `SubmitSnapshotRequest` and this in a corner case can cause launching the  `generateSnapshot` goroutine for the same snapshot twice.

I think that it would be much easier to understand the idea of post commit snapshot generation, if you don't have this function altogether and simply move the code at the end of commit function. There, you already have the block number and moreover, this function is not specific to snapshot mgmt as such.